ospf6d/ospf6_gr.c

```diff
@@ -0,0 +1,746 @@
+
+			snprintfrr(reason, sizeof(reason),
+				   "detected inconsistent LSA %s [area %pI4]",
+				   lsa->name, &area->area_id);
+			ospf6_gr_restart_exit(ospf6, reason);
+			return;
+		}
+	}
+}
+
+/* Check if there's a fully formed adjacency with the given neighbor ID. */
+static bool ospf6_gr_check_adj_id(struct ospf6_area *area,
+				  in_addr_t neighbor_router_id)
+{
+	struct ospf6_neighbor *nbr;
+
+	nbr = ospf6_area_neighbor_lookup(area, neighbor_router_id);
+	if (!nbr || nbr->state < OSPF6_NEIGHBOR_FULL) {
+		if (IS_DEBUG_OSPF6_GR_HELPER)
+			zlog_debug("GR: missing adjacency to router %pI4",
+				   &neighbor_router_id);
+		return false;
+	}
+
+	return true;
+}
+
+static bool ospf6_gr_check_adjs_lsa_transit(struct ospf6_area *area,
+					    in_addr_t neighbor_router_id,
+					    uint32_t neighbor_interface_id)
+{
+	struct ospf6 *ospf6 = area->ospf6;
+
+	/* Check if we are the DR. */
+	if (neighbor_router_id == ospf6->router_id) {
+		struct ospf6_lsa *lsa;
+		char *start, *end, *current;
+		struct ospf6_network_lsa *network_lsa;
+		struct ospf6_network_lsdesc *lsdesc;
+
+		/* Lookup Network LSA corresponding to this interface. */
+		lsa = ospf6_lsdb_lookup(htons(OSPF6_LSTYPE_NETWORK),
+					neighbor_interface_id,
+					neighbor_router_id, area->lsdb);
+		if (!lsa)
+			return false;
+
+		/* Iterate over all routers present in the network. */
+		network_lsa = (struct ospf6_network_lsa
+				       *)((char *)lsa->header
+					  + sizeof(struct ospf6_lsa_header));
+		start = (char *)network_lsa + sizeof(struct ospf6_network_lsa);
+		end = (char *)lsa->header + ntohs(lsa->header->length);
+		for (current = start;
+		     current + sizeof(struct ospf6_network_lsdesc) <= end;
+		     current += sizeof(struct ospf6_network_lsdesc)) {
+			lsdesc = (struct ospf6_network_lsdesc *)current;
+
+			/* Skip self in the pseudonode. */
+			if (lsdesc->router_id == ospf6->router_id)
+				continue;
+
+			/*
+			 * Check if there's a fully formed adjacency with this
+			 * router.
+			 */
+			if (!ospf6_gr_check_adj_id(area, lsdesc->router_id))
+				return false;
+		}
+	} else {
+		struct ospf6_neighbor *nbr;
+
+		/* Check if there's a fully formed adjacency with the DR. */
+		nbr = ospf6_area_neighbor_lookup(area, neighbor_router_id);
+		if (!nbr || nbr->state < OSPF6_NEIGHBOR_FULL) {
+			if (IS_DEBUG_OSPF6_GR_HELPER)
+				zlog_debug(
+					"GR: missing adjacency to DR router %pI4",
+					&neighbor_router_id);
+			return false;
+		}
+	}
+
+	return true;
+}
+
+static bool ospf6_gr_check_adjs_lsa(struct ospf6_area *area,
+				    struct ospf6_lsa *lsa)
+{
+	struct ospf6_router_lsa *router_lsa;
+	char *start, *end, *current;
+
+	router_lsa =
+		(struct ospf6_router_lsa *)((char *)lsa->header
+					    + sizeof(struct ospf6_lsa_header));
+
+	start = (char *)router_lsa + sizeof(struct ospf6_router_lsa);
+	end = (char *)lsa->header + ntohs(lsa->header->length);
+	for (current = start;
+	     current + sizeof(struct ospf6_router_lsdesc) <= end;
```

Comment:
This seems odd ... should there be a log or assert or something here? I don't see what this line of code is doing.

Reply:
What this loop does is to iterate over all interfaces of the given Router-LSA. It isn't pretty, but that's how it's done in other parts of the daemon as well.

I'll add a comment to clarify things a bit. Later on, we might consider adding an iteration macro to simplify all places where this iteration needs to be done.